Common/AR/source/LocationDataSource.cpp

```diff
@@ -179,24 +235,89 @@ QCompass* LocationDataSource::compass() const
 {
-  if (m_compass == compass || !compass)
+  if (m_isStarted || m_compass == compass)
     return;
 
   m_compass = compass;
+  emit compassChanged();
+}
 
+/*!
+  \internal
+  Connect or disconnect the geoPositionSource and compass signals.
+  If locationTrackingMode is "Continuous", the connection still alive until stopped.
+  If locationTrackingMode is "Initial", the connection is disconnected after the first signal
+  is emitted.
+ */
+void LocationDataSource::updateObjectsAndConnections()
+{
+  // The geoPositionSource and compass objects can be created by the user and
+  // can be still alive when the pointers are set to null. Then the connections
+  // must be disconnected manually.
+  disconnect(m_geoPositionSourceConnection);
   disconnect(m_compassConnection);
-  m_compassConnection = connect(m_compass, &QCompass::readingChanged, this, [this]()
-  {
-    // emit the new heading if available
-    QCompassReading* reading = m_compass->reading();
-    Q_CHECK_PTR(reading);
-    emit headingChanged(reading->azimuth());
-  });
 
-  emit compassChanged();
+  // Create a default geo position source and compass if necessary.
+  if (!m_geoPositionSource)
+    setGeoPositionSource(QGeoPositionInfoSource::createDefaultSource(this));
+
+  if (!m_compass)
+    setCompass(new QCompass(this));
+
+  // Connect the change signals.
+  if (m_locationTrackingMode == LocationTrackingMode::Continuous)
+  {
+    // Connect the geoPositionSource.
+    m_geoPositionSourceConnection = connect(m_geoPositionSource, &QGeoPositionInfoSource::positionUpdated,
+                                            this, [this](const QGeoPositionInfo &positionInfo)
+    {
+      // Emit the new position if available.
+      const QGeoCoordinate& coordinate = positionInfo.coordinate();
+      if (coordinate.isValid())
+        emit locationChanged(coordinate.latitude(), coordinate.longitude(), coordinate.altitude());
+    });
+
+    // Connect the geoPositionSource.
+    m_compassConnection = connect(m_compass, &QCompass::readingChanged, this, [this]()
+    {
+      // emit the new heading if available
+      QCompassReading* reading = m_compass->reading();
+      Q_CHECK_PTR(reading);
+      emit headingChanged(reading->azimuth());
+    });
+  }
+  else if (m_locationTrackingMode == LocationTrackingMode::Initial)
+  {
+    // Connect the geoPositionSource and disconnect when the first signal is emitted.
+    m_geoPositionSourceConnection = connect(m_geoPositionSource, &QGeoPositionInfoSource::positionUpdated,
+                                            this, [this](const QGeoPositionInfo &positionInfo)
+    {
+      // Emit the new position if available.
+      const QGeoCoordinate& coordinate = positionInfo.coordinate();
+      if (coordinate.isValid())
+        emit locationChanged(coordinate.latitude(), coordinate.longitude(), coordinate.altitude());
+
+      // Disconnect the signal.
+      disconnect(m_compassConnection);
```

Comment:
I think it would simplify the code if we can use the same lambda logic for both `LocationTrackingMode::Initial` and `LocationTrackingMode::Continuous`.

the only difference would be:
```c++
if (m_locationTrackingMode == LocationTrackingMode::Initial)
  disconnect(m_compassConnection);
```

I don't know if that's worth the never-ending runtime check for Continuous mode. What do you think?